sql/expression/bit_ops.go

```diff
@@ -0,0 +1,429 @@
+}
+
+func (b *BitOp) DebugString() string {
+	return fmt.Sprintf("(%s %s %s)", sql.DebugString(b.Left), b.Op, sql.DebugString(b.Right))
+}
+
+// IsNullable implements the sql.Expression interface.
+func (b *BitOp) IsNullable() bool {
+	return b.BinaryExpression.IsNullable()
+}
+
+// Type returns the greatest type for given operation.
+func (b *BitOp) Type() sql.Type {
+	rTyp := b.Right.Type()
+	if sql.IsDeferredType(rTyp) {
+		return rTyp
+	}
+	lTyp := b.Left.Type()
+	if sql.IsDeferredType(lTyp) {
+		return lTyp
+	}
+
+	if sql.IsUnsigned(lTyp) && sql.IsUnsigned(rTyp) {
+		return sql.Uint64
+	} else if sql.IsSigned(lTyp) && sql.IsSigned(rTyp) {
+		return sql.Int64
+	}
+	return sql.Float64
+}
+
+// WithChildren implements the Expression interface.
+func (b *BitOp) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	if len(children) != 2 {
+		return nil, sql.ErrInvalidChildrenNumber.New(b, len(children), 2)
+	}
+	return NewBitOp(children[0], children[1], b.Op), nil
+}
+
+// Eval implements the Expression interface.
+func (b *BitOp) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	lval, rval, err := b.evalLeftRight(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if lval == nil || rval == nil {
+		return nil, nil
+	}
+
+	lval, rval, err = b.convertLeftRight(ctx, lval, rval)
+	if err != nil {
+		return nil, err
+	}
+
+	switch strings.ToLower(b.Op) {
+	case sqlparser.BitAndStr:
+		return bitAnd(lval, rval)
+	case sqlparser.BitOrStr:
+		return bitOr(lval, rval)
+	case sqlparser.BitXorStr:
+		return bitXor(lval, rval)
+	case sqlparser.ShiftLeftStr:
+		return shiftLeft(lval, rval)
+	case sqlparser.ShiftRightStr:
+		return shiftRight(lval, rval)
+	}
+
+	return nil, errUnableToEval.New(lval, b.Op, rval)
+}
+
+func (b *BitOp) evalLeftRight(ctx *sql.Context, row sql.Row) (interface{}, interface{}, error) {
+	var lval, rval interface{}
+	var err error
+
+	if i, ok := b.Left.(*Interval); ok {
+		lval, err = i.EvalDelta(ctx, row)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		lval, err = b.Left.Eval(ctx, row)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if i, ok := b.Right.(*Interval); ok {
+		rval, err = i.EvalDelta(ctx, row)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		rval, err = b.Right.Eval(ctx, row)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return lval, rval, nil
+}
+
+func (b *BitOp) convertLeftRight(ctx *sql.Context, left interface{}, right interface{}) (interface{}, interface{}, error) {
+	var err error
+
+	typ := b.Type()
+
+	left, err = typ.Convert(left)
+	if err != nil {
+		ctx.Session.Warn(&sql.Warning{
+			Level:   "Warning",
+			Code:    mysql.ERTruncatedWrongValue,
+			Message: fmt.Sprintf("Truncated incorrect %s value: '%v'", typ.String(), left),
+		})
+		// the value is interpreted as 0, but we need to match the type of the other valid value
+		// to avoid additional conversion, the nil value is handled in each operation
+		left = nil
+	}
```

Comment:
(minor) It might be worth thinking about extracting this to a helper method. It seems like we have this pattern twice in each `convertLeftRight` function. Looks like the `convertLeftRight` functions have some slight differences, but we might still be able to pull out a smaller portion that does the conversion check, sets a warning for truncation. Not a huge deal, but might tidy up the code a little bit.